src/middlewares/restrict-user-details.js

```diff
@@ -0,0 +1,102 @@
+module.exports = (config, { strapi }) => {
+  return async (ctx, next) => {
+    await next();
+    try {
+      if (ctx.request.method === "GET" && ctx.request.url.includes("/api/")) {
+        const handler = ctx.state.route.handler;
+        const user = ctx.state.user;
+        const idx = handler.lastIndexOf(".");
+        const modelName = handler.substring(0, idx);
+        const action = handler.substring(idx + 1);
+        const metadata = strapi.db.metadata;
+
+        const fieldsRestricter = new FieldsRestricter(metadata, config);
+        
+        ctx.response.body = fieldsRestricter.restrictFields(ctx.response.body, modelName, action, user);
+      }
+    } catch (error) {
+      console.error(error);
+    }
+  };
+};
+
+class FieldsRestricter {
+  constructor(metadata, config) {
+    this.metadata = metadata;
+    this.config = config;
+  }
+
+  findAndRestrictUserDetails(item, modelName) {
+    if (modelName === "user") {
+      return this.restrictNonWrappedItem(item, this.config.allowedUserFields);
+    }
+    if (modelName === "plugin::users-permissions.user") {
```

Comment:
When will we see this modelName instead of just `user`?

Reply:
When we call `/api/users/1` we will get handler `user.findOne`, but let say we call `api/idea-cards/1?populate=ideaOwner`, that ideaOwner will be a relation type and will target not `user` but  `plugin::users-permissions.user`. Other modelNames are more consistent `api::profile.profile` is the same in both handler and relation target type.